openapi-cli/src/main/java/io/ballerina/openapi/generators/schema/BallerinaSchemaGenerator.java

```diff
@@ -228,29 +231,45 @@ private void generateTypeDefinitionNode(List<TypeDefinitionNode> typeDefinitionN
-                    createToken(SyntaxKind.RECORD_KEYWORD), createToken(OPEN_BRACE_TOKEN),
-                    fieldNodes, null, createToken(SyntaxKind.CLOSE_BRACE_TOKEN));
-            MarkdownDocumentationNode documentationNode = createMarkdownDocumentationNode(createNodeList(schemaDoc));
-            MetadataNode metadataNode = createMetadataNode(documentationNode, createNodeList(typeAnnotations));
-            typeDefNode = NodeFactory.createTypeDefinitionNode(metadataNode,
-                    createToken(PUBLIC_KEYWORD),
-                    createToken(TYPE_KEYWORD),
-                    typeName, recordTypeDescriptorNode, createToken(SEMICOLON_TOKEN));
-            typeDefinitionNodeList.add(typeDefNode);
-
+            Schema schema = this.schemas.get(typeRef.text());
+            String schemaType = schema.getType();
+            if ((schemaType != null && !schemaType.equals(ARRAY_TYPE)) || (schemaType == null &&
+                    schema.get$ref() != null)) {
+                MarkdownDocumentationNode documentationNode = createMarkdownDocumentationNode(
+                        createNodeList(schemaDoc));
+                MetadataNode metadataNode = createMetadataNode(documentationNode, createNodeList(typeAnnotations));
+                String typeDescriptorName = typeRef.text();
+                typeDescriptorName = getNullableType(schemaValue, typeDescriptorName);
+                typeDefNode = getTypeDefinitionNode(createIdentifierToken(typeDescriptorName),
+                        getValidName((defineDatatypeName), true),
+                        metadataNode);
+                typeDefinitionNodeList.add(typeDefNode);
+            } else {
+                TypeReferenceNode recordField = NodeFactory.createTypeReferenceNode(createToken(ASTERISK_TOKEN),
+                        typeRef,
+                        createToken(SEMICOLON_TOKEN));
+                recordFieldList.add(recordField);
+                NodeList<Node> fieldNodes = AbstractNodeFactory.createNodeList(recordFieldList);
+                RecordTypeDescriptorNode recordTypeDescriptorNode = NodeFactory.createRecordTypeDescriptorNode(
+                        createToken(SyntaxKind.RECORD_KEYWORD), createToken(OPEN_BRACE_TOKEN),
+                        fieldNodes, null, createToken(SyntaxKind.CLOSE_BRACE_TOKEN));
+                MarkdownDocumentationNode documentationNode = createMarkdownDocumentationNode(
+                        createNodeList(schemaDoc));
+                MetadataNode metadataNode = createMetadataNode(documentationNode, createNodeList(typeAnnotations));
+                typeDefNode = NodeFactory.createTypeDefinitionNode(metadataNode,
+                        createToken(PUBLIC_KEYWORD),
+                        createToken(TYPE_KEYWORD),
+                        typeName, recordTypeDescriptorNode, createToken(SEMICOLON_TOKEN));
+                typeDefinitionNodeList.add(typeDefNode);
```

Comment:
```suggestion
                String typeDescriptorName = getValidName(extractReferenceType(schemaValue.get$ref()), true);
                Schema<?> refSchema = openAPI.getComponents().getSchemas().get(typeDescriptorName);
                typeDescriptorName = getNullableType(refSchema, typeDescriptorName);
                MarkdownDocumentationNode documentationNode = createMarkdownDocumentationNode(createNodeList(schemaDoc));
                MetadataNode metadataNode = createMetadataNode(documentationNode, createNodeList(typeAnnotations));
                typeDefNode = getTypeDefinitionNode(createIdentifierToken(typeDescriptorName),
                        getValidName((defineDatatypeName), true),
                        metadataNode);
```
We don't need to specify two approaches for primitive types and object types when one schema is directly referring to another schema.

Reply:
These changes were added only to fix the simple record reference issue. However, we can remove complete record inclusions for the types that have only the referred field but no other specific fields as you suggested in your example.

Reply:
Added the suggested improvement with a test case